### DISCUSSION TITLE
Biomass heating network, biomass composition.

### DISCUSSION THREAD

**Discussion by RubenHeurckmans** [oemof/tespy#491] [Q&A]
Hi, For my thesis I have to reproduce a network from a paper in which heat pumps are integrated into a biomass heating network to recover heat from the exhaust gases. Is it possible to add my own biomass composition as fuel?
Thanks

**Comment by fwitte** [comment]
Hi @RubenHeurckmans,

you can specify different compositions for the combustion chamber or combustion engine classes. There are a couple of gases which are registered as combustion gases ('methane', 'ethane', 'propane', 'butane', 'hydrogen', 'nDodecane'). You can mix them with each other or also with CO2 or water etc.. Would that work?

Best

**Reply by RubenHeurckmans** [reply]
Hi,
Thanks for your answer, sorry for my late response. I woud now just like to add a typical biomass approximation: CH(1,44)O(0,66). Is that possible? 
Thanks,
Ruben

**Reply by fwitte** [reply]
Can you provide a figure or something to outline your concept. I am not sure, if I understood correctly... What flows of materials and energy are included and what components. Thanks

**Reply by RubenHeurckmans** [reply]
Since biomass used to be cheap, many old power stations are not efficient. Nowadays the price is rising and it is important to improve efficiency. For my modest thesis I base myself on 2 previously published papers in which a heat pump is used to recover heat from the flue gases.

Using Tespy on Python, I have to calculate how and to what extent the addition of that heat pump adds value to improving efficiency.

I intended to simply represent the boiler as a heat exchanger in which I could input a heat flow using the LHV and the mass flow rate of my biomass. But I now realize that I don't really know how I can use an air preheater, nor how I can use the temperature of my flue gases to draw conclusions.

I hope I'm not asking too much of you, but I may not have been working on the subject long enough and may not be asking sensible questions in that way.

Attached you will find 2 screenshots of the type of network I need to reproduce.

Thanks in advance,

Ruben
![NetwerkThesis2](https://github.com/oemof/tespy/assets/160655085/e0aa377a-d5a6-4708-a90e-0316b1dc3aff)
![NetwerkThesis](https://github.com/oemof/tespy/assets/160655085/71283f6d-3df5-4fa8-931f-1d0c56dbc959)

**Reply by fwitte** [reply]
> Thanks for your answer, sorry for my late response. I woud now just like to add a typical biomass approximation: CH(1,44)O(0,66). Is that possible?

Okay, you want to have a mixture of CH4 with 44 % and CO2 with 56 % (what is CH1,44 and O 0,66?)? You could have a heat exchanger attached to a combustion chamber:

```python
from tespy.networks import Network
from tespy.components import HeatExchanger, Source, Sink, Valve, HeatExchangerSimple, CycleCloser, Compressor, Turbine, CombustionChamber
from tespy.connections import Connection, Bus
from CoolProp.CoolProp import PropsSI
from tespy.tools.helpers import get_chem_ex_lib


nwk = Network(p_unit="bar", T_unit="C")

biogas = Source("biogas")
air = Source("air")
cc = CombustionChamber("combustion")
he = HeatExchanger("boiler")

fg = Sink("flue gas")

water_in = Source("water cold")
water_out = Sink("water hot")

c1 = Connection(biogas, "out1", cc, "in1", label="1")
c2 = Connection(air, "out1", cc, "in2", label="2")
c3 = Connection(cc, "out1", fg, "in1", label="3")
# c4 = Connection(biogas, "out1", cc, "in1", label="4")

nwk.add_conns(c1, c2, c3)

c1.set_attr(fluid={"CH4": 0.44, "CO2": 0.56}, m=1, T=25, p=1)
c2.set_attr(fluid={"N2": 0.77, "O2": 0.23}, m=20, T=25)

nwk.solve("design")

nwk.print_results()

nwk.del_conns(c3)

c3 = Connection(cc, "out1", he, "in1", label="3")
c4 = Connection(he, "out1", fg, "in1", label="4")

c11 = Connection(water_in, "out1", he, "in2", label="11")
c12 = Connection(he, "out2", water_out, "in1", label="12")

nwk.add_conns(c3, c4, c11, c12)

c4.set_attr(T=100, p=1)
c11.set_attr(p=1, T=50, fluid={"water": 1})
c12.set_attr(p=1, T=75)

nwk.solve("design")

nwk.print_results()
```

If you want to have an integrated combustion chamber and boiler, you would have to create your own component for that.

**Reply by fwitte** [reply]
As a hint: Build you system step by step as I did. Add one component at a time, it will make the debugging much easier. You can adapt the parameter obviously to what you want to have in your flowsheet.

**Reply by RubenHeurckmans** [reply]
It is biomass, like wood pellets for example, not biogas. The notation "CH1.44O0.66" represents the relative molar fractions of carbon (C), hydrogen (H), and oxygen (O) in biomass. The number following each element indicates its molar fraction. So, for example, "CH1.44O0.66" means that for each unit of biomass, there are 1.44 moles of carbon, 1 mole of hydrogen, and 0.66 moles of oxygen. This is a simplified way to describe the composition of biomass without delving into the specific molecular structure of the compounds present in the biomass.

**Reply by RubenHeurckmans** [reply]
But also, thanks for your help ofcourse!

**Reply by fwitte** [reply]
Ah okay, got it, thank you for your patience. Then I think, this is not yet implemented in tespy. You are welcome to integrate it. This requires two steps and a little bit of digging into tespy's source code:

- creating a new fluid, that describes the thermo-phyisical properties of the biomass, that would be temperature-pressure-enthalpy relations mostly.
- adding calculation methods for the combustion to be able to calculate the specific lower heating value of the fluid.

If you are interested, please let me know. We can have a short meeting and I can introduce you, where to start with the lower heating value calculation. For the definition of custom fluids, you can either have a look at CoolProp's incompressible back-end (I think there should be some information on how to integrate custom fluids) or you can simply have a couple of basic equations to describe the properties and put them into your own wrapper, like is done [here]( https://tespy.readthedocs.io/en/main/modules/fluid_properties.html#implementing-a-custom-engine) or [here](https://github.com/IPUdk/tespy/tree/EEMAP/dev/incompressiblesTests/mywrapper) with the definition of the `CustomWrapper` [here](https://github.com/IPUdk/tespy/blob/EEMAP/dev/src/tespy/tools/fluid_properties/CustomWrapper.py).

Best
